pkg/lvgl: add initial support for LittlevGL to RIOT

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds support to the [LittlevGL](https://github.com/littlevgl/lvgl) package to RIOT. I've been looking to it for quite some time and now that the ili9341 driver is in, I think it's time to move forward with this.

This PR also introduces a generic (and still basic) API for display devices, it follows the same scheme as netdev.

There's also a test application that display a system monitor on a screen. It was tested on the stm32f429i-disc1 board.

I know other people in the community (@bergzand, @fjmolinas) have already done some work with this package and maybe they can provide some good advice here so we can move forward faster.

Ideas for future improvements:
- improve package configuration macros: maybe even in this PR (I think it's the weakest point of this PR)
- provide a driver for the stm32f429i-disco touch screen, some months ago I started to write a basic driver but it still needs some work
- provide support for [lvgl-examples](https://github.com/littlevgl/lv_examples) in another package (that depends on this one), this way it will be simpler to try different lvgl features.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Grab an stm32f429i-disc1 board and run the following command
```
make -C tests/pkg_lvgl flash term
```

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Based on ##13262

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,87 @@
+
+    while (1) {
+        lv_task_handler();
+        xtimer_usleep(5 * US_PER_MS);
```
does this need to be called 200 times per second?

kaspar030 replied:
Seems so: https://docs.littlevgl.com/en/html/porting/task-handler.html

kaspar030 replied:
There's ways around: https://docs.littlevgl.com/en/html/porting/sleep.html
(we can improve that later)

aabadie replied:
Thanks, I'll give this a try.

aabadie replied:
> we can improve that later

I can't think of an easy solution for this or without having to patch littlevgl. Maybe I'm missing something.

bergzand commented on pkg/lvgl/include/lv_conf.h:
```diff
@@ -0,0 +1,597 @@
+#if LV_TICK_CUSTOM == 1
+#define LV_TICK_CUSTOM_INCLUDE  "something.h"       /*Header for the sys time function*/
+#define LV_TICK_CUSTOM_SYS_TIME_EXPR (millis())     /*Expression evaluating to current systime in ms*/
+#endif   /*LV_TICK_CUSTOM*/
```
If you modify this to use `xtimer.h` and `xtimer_now_usec() / US_PER_MS` you don't need to call the `lv_tick_inc()` periodically. (or did you already solve that somewhere else?)

aabadie replied:
Thanks a lot, I think this explains why I have problem with the actual port :)

aabadie replied:
Ok, I changed that and now it works like a charm: the system monitor is refreshed correctly :)
(I spent quite some time on this problem and couldn't make it work...)

bergzand commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,87 @@
+
+    while (1) {
+        lv_task_handler();
+        xtimer_usleep(5 * US_PER_MS);
```
Wouldn't it be cleaner to use `xtimer_periodic_sleep()` for this?

aabadie replied:
This function doesn't exist ;) I guess you meant `xtimer_periodic_wakeup`.

bergzand commented on tests/pkg_lvgl/Makefile:
```diff
@@ -0,0 +1,13 @@
+USEMODULE += xtimer
+
+USEMODULE += ili9341
+CFLAGS += -DILI9341_LE_MODE
```
You can remove this define if you let LittleVGL switch the endianness of the buffer (with `LV_COLOR_16_SWAP`). That significantly speeds up the SPI transfer throughput since the ili9341 doesn't have to swap every `uint16_t` before transfering allowing the SPI peripheral to fully utilize DMA for the whole buffer transfer.

kaspar030 commented:
I tested this in the PineTime. Seems to work, with just minor modifications!

1. the pinetime has only 240x240 pixels. needed to adjust that. I think this needs to become a parameter if the ili9341, and I think the disp_dev needs a way to query that.

2. the pinetime uses gpios to control the backlight. Maybe support for that also needs to be added to disp_dev?

I have a picture, but somehow my Gnome/Wayland doesn't let me paste it here. ;(

aabadie commented:
>  I think this needs to become a parameter if the ili9341, and I think the disp_dev needs a way to query that

It's not that simple without patching lvgl. The `LV_HOR_RES_MAX` and `LV_VER_RES_MAX` are used internally when initializing the display, see [here](https://github.com/littlevgl/lvgl/blob/master/src/lv_hal/lv_hal_disp.c#L61).
in `lv_conf.h`, these macro can be overridden: this provides compile time configuration.

> Maybe support for that also needs to be added to disp_dev?

Probably doable, with get/set functions.

kaspar030 commented:
> It's not that simple without patching lvgl. The `LV_HOR_RES_MAX` and `LV_VER_RES_MAX` are used internally when initializing the display, see [here](https://github.com/littlevgl/lvgl/blob/master/src/lv_hal/lv_hal_disp.c#L61).
> in `lv_conf.h`, these macro can be overridden: this provides compile time configuration.

Hm, looking at the "porting" example [here](https://github.com/littlevgl/lvgl/blob/51928059d2dda465ab76ee66cf37b9d2edb9f64e/porting/lv_port_disp_template.c#L92-L103), they first initialize the driver struct (using `lv_disp_drv_init()`, which just clears and sets values), then afterwards override the resolution and buffers. Maybe we can do it similarly?

BTW, at this point, does it make sense to introduce `disp_dev_t`? We could just use `lv_disp_t`. It even provides a method to add user defined fields (https://github.com/littlevgl/lvgl/blob/51928059d2dda465ab76ee66cf37b9d2edb9f64e/src/lv_hal/lv_hal_disp.h#L115-L117).

bergzand commented:
I see the merits of having a `display_dev` abstraction for display devices. However for now I would like to propose to either:
 - split the `display_dev` header to a separate PR and continue this with only support for the ili9341 device driver, it is the only display supported at the moment anyway.

or:
 - Keep the `display_dev` as it is now but mark it as experimental, I think it is a good start and we can iterate over it in a follow up PR.

aabadie commented:
> I see the merits of having a display_dev abstraction for display devices

The abstraction design has no external dependency so each display driver can provide the required driver functions (only one for the moment). If we reuse the abstraction API that comes from lvgl, then this introduce a hard dependency to it. I would prefer to avoid this.
@fjmolinas has a driver for other LCD modules that could benefit from this very easily.

So for now I would use option B, as proposed by @bergzand: keep the disp_dev header and improve later.

Regarding the while loop in the task thread, I agree it's not nice, especially because of power consumption.
We could use a blocking `msg_receive` function there when the system is inactive and send a message to this thread on wakeup. Something like:

```c
while (1) {
    /*Normal operation (no sleep) in < 1 sec inactivity*/
    if (lv_disp_get_inactive_time(NULL) < 1000) {
        lv_task_handler();
    }
    /* Block after 1 sec inactivity */
    else {
        msg_t msg;
        msg_receive(&msg);
    }
    xtimer_usleep(5 * US_PER_MS);
}
```

Then the screen could be explicitly restarted from user code with a simple `msg_send`.

kaspar030 commented:
> So for now I would use option B, as proposed by @bergzand: keep the disp_dev header and improve later.

ok

kaspar030 commented:
> Then the screen could be explicitly restarted from user code with a simple `msg_send`.

better use thread flags. Cannot get lost (e.g., if set from ISR), doesn't block caller.

aabadie commented:
> better use thread flags.

@kaspar030, I just pushed a couple of commits that implement a mechanism to allow stopping the lvgl task handler thread after some inactivity period (defined by the LVGL_ACTIVITY_PERIOD macro). The glue code now provides an extra function `lvgl_wakeup` that needs to be called when lvgl should be restarted (or the inactivity limit be kicked).

In the sysmon example, this function is called each time the sysmon_task is called: this ensures the chart is always refreshed. But if you tweak the REFR_TIME macro in main.c to be greater than LVGL_ACTIVITY_PERIOD, then lvgl is stopped after LVGL_ACTIVITY_PERIOD is elapsed and an explicit call to lvgl_wakeup is needed to restart it (this can be done in a gpio callback for example).

What do you think of the approach ?

aabadie commented:
I added a bit of documentation.

@kaspar030, if you think the thread_flags is good enough, this PR is close to be mergeable I think.

kaspar030 commented on pkg/lvgl/Makefile:
```diff
@@ -0,0 +1,24 @@
+all: $(LVGL_MODULES)
+	"$(MAKE)" -C $(PKG_BUILDDIR) -f $(CURDIR)/Makefile.$(PKG_NAME)
+
+lvgl_%:
```
Seeing that the module name is available here, how about:
```
	MODULE=$@ "$(MAKE)" -C $(PKG_BUILDDIR)/src/lv_$* -f $(CURDIR)/Makefile.module
```

That way we only need one (instead of 7) ```Makefile.*```.

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,117 @@
+#define LVGL_TASK_HANDLER_DELAY (5 * US_PER_MS)
+#endif
+
+static char _task_thread_stack[THREAD_STACKSIZE_MAIN];
```
This should be "THREAD_STACKSIZE_DEFAULT" (or maybe _MEDIUM or _LARGE?).
A user might have a large main and thus overrides "THREAD_STACKSIZE_MAIN", which shouldn't affect lvgl's thread.

aabadie replied:
Changed to THREAD_STACKSIZE_LARGE, otherwise it doesnt fit.

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,117 @@
+static lv_color_t buf[LVGL_COLOR_BUF_SIZE];
+static disp_dev_t *_dev = NULL;
+
+void *_task_thread(void *arg)
```
static?

aabadie replied:
all fixed

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,117 @@
+    return NULL;
+}
+
+void _disp_map(lv_disp_drv_t *drv, const lv_area_t *area, lv_color_t *color_p)
```
static?

kaspar030 commented on tests/pkg_lvgl/Makefile:
```diff
@@ -0,0 +1,11 @@
+
+USEPKG += lvgl
+USEMODULE += ili9341
+USEMODULE += xtimer
```
lvgl already depends on xtimer?

aabadie replied:
a left-over from previous versions, removed.

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,124 @@
+#include "disp_dev.h"
+
+#ifndef LVGL_TASK_THREAD_PRIO
+#define LVGL_TASK_THREAD_PRIO   (THREAD_PRIORITY_MAIN + 1)
```
so, lower priority (higher value) than main?

aabadie replied:
I tried both versions and it works the same. Not sure which one is best.

kaspar030 replied:
ok. I guess it depends on what's actually running in main.

kaspar030 commented:
Please rebase. And maybe squash, might make reviewing easier.

aabadie commented:
Squashed and rebased.

Travis is already reporting an issue with Doxygen:
> Included by graph for 'periph_cpu.h' not generated, too many nodes. Consider increasing DOT_GRAPH_MAX_NODES.

Not sure how to handle this.

kaspar030 commented:
> Not sure how to handle this.

> > Consider increasing DOT_GRAPH_MAX_NODES.

we should consider this. Is the output somehow large? Why does this pop up now...

aabadie commented:
> Why does this pop up now...

Bad luck :)

aabadie commented:
It's also happening in another PR: https://ci.riot-os.org/RIOT-OS/RIOT/13480/b27ae335af180ca0b2b3f3b44103e2b9dca0b9ac/output/static_tests.txt

bergzand commented on pkg/lvgl/Makefile:
```diff
@@ -0,0 +1,24 @@
+PKG_NAME=lvgl
+PKG_URL=https://github.com/littlevgl/lvgl
+PKG_VERSION=v6.1.1
```
This should be a commit hash with the version as a comment

aabadie replied:
IMHO, that doesn't make any difference

bergzand commented on pkg/lvgl/Makefile:
```diff
@@ -0,0 +1,24 @@
+PKG_NAME=lvgl
+PKG_URL=https://github.com/littlevgl/lvgl
+PKG_VERSION=v6.1.1
```
Version 6.1.2 was released 7 days ago and only includes bug fixes. Maybe you could bump it if that's okay with you.

aabadie replied:
IMHO, that doesn't make any difference

bergzand commented on pkg/lvgl/Makefile.dep:
```diff
@@ -0,0 +1,18 @@
+USEMODULE += lvgl_objx
+USEMODULE += lvgl_themes
+
+USEMODULE += lvgl_contrib
```
I would strongly prefer to have this optional. In case a user wants to handle the LVGL thread himself.

aabadie replied:
Ok, done!

bergzand commented on pkg/lvgl/Makefile.dep:
```diff
@@ -0,0 +1,18 @@
+USEMODULE += lvgl_contrib
+USEMODULE += disp_dev
+USEMODULE += core_thread_flags
+USEMODULE += xtimer
```
These should be a dependency of lvgl_contrib and not of the lvgl package itself I think.

aabadie replied:
Done

aabadie commented:
I think I addressed all your comments @bergzand :)

bergzand commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,124 @@
+#endif
+
+#ifndef LVGL_THREAD_FLAG
+#define LVGL_THREAD_FLAG        (0x4242)
```
Thread flags are stored as a bit mask, not as a number. Shouldn't this be something like `(1 << FLAG_NUM)`?

bergzand commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,124 @@
+
+    disp_dev_map(_dev,
+                 (uint16_t)area->x1, (uint16_t)area->x2,
+                 (uint16_t)area->y1, (uint16_t)area->y2,
```
Are these casts really necessary?

aabadie commented:
Comments addressed @bergzand :)

aabadie commented:
ping @bergzand :)

kaspar030 commented on pkg/lvgl/contrib/lvgl.c:
```diff
@@ -0,0 +1,122 @@
+                                     LVGL_TASK_THREAD_PRIO, THREAD_CREATE_STACKTEST,
+                                     _task_thread, NULL, "_task_thread");
+
+    xtimer_sleep(1);
```
why the one second sleep?

aabadie replied:
Without it crashes on adafruit-clue and maybe also on the pinetime but I haven't checked this one.

aabadie replied:
Ok, I found the problem and could fix it :)

kaspar030 commented on tests/pkg_lvgl/main.c:
```diff
@@ -0,0 +1,152 @@
+
+    /* Force a wakeup of lvgl when each task is called: this ensure an activity
+       is triggered and wakes up lvgl during the next LVGL_ACTIVITY_PERIOD ms */
+    lvgl_wakeup();
```
shouldn't lvgl be woken up *after* queueing changes?

aabadie replied:
I moved it after.

bergzand commented on tests/pkg_lvgl/main.c:
```diff
@@ -0,0 +1,152 @@
+                       (int)mem_mon.total_size - mem_mon.free_size,
+                       mem_mon.free_size, mem_mon.frag_pct);
+
+    lv_label_set_text(info_label, buf_long);
```
Why not use `lv_label_set_fmt` here and skip the two `lv_snprintf` calls?

bergzand commented:
Please squash after addressing my last comment

aabadie commented:
Awesome ! Thanks a lot for the reviews and tests :)
